yt/frontends/amrvac/data_structures.py

```diff
@@ -250,16 +251,101 @@ def _set_code_unit_attributes(self):
 
-        setdefaultattr(self, "length_unit", self.quan(1, "cm"))
-        setdefaultattr(self, "mass_unit", self.quan(1, "g"))
-        setdefaultattr(self, "time_unit", self.quan(1, "s"))
+        # note: normalisations in AMRVAC have 3 degrees of freedom. The user can specify a unit length and unit
+        # numberdensity, with the third option either a unit temperature OR unit velocity.
+        # If unit temperature is specified then unit velocity will be calculated accordingly, and vice-versa.
+        # AMRVAC does not allow to specify any other normalisation beside those four.
+
+        # note: _override_code_units() has already been called, so self.units_override has been set.
+        # _override_code_units() sets supplied items in units_override as attributes.
+        # Handled items are length, time, mass, velocity, magnetic, temperature
+
+        # note: yt sets hydrogen mass equal to proton mass, amrvac doesn't.
+        mp_cgs = self.quan(1.672621898e-24, 'g')  # This value is taken from AstroPy
+        He_abundance = 0.1  # hardcoded parameter in AMRVAC
+
+        # get self.length_unit if overrides are supplied, otherwise use default
+        length_unit = getattr(self, 'length_unit', self.quan(1, 'cm'))
+
+        # 1. calculations for mass, density, numberdensity
+        if 'mass_unit' in self.units_override:
+            # in this case unit_mass is supplied (and has been set as attribute)
+            mass_unit = self.mass_unit
+            density_unit = mass_unit / length_unit**3
+            numberdensity_unit = density_unit / ((1.0 + 4.0 * He_abundance) * mp_cgs)
+        else:
+            # other case: numberdensity is supplied. Fall back to one (default) if no overrides supplied
+            numberdensity_override = self.units_override.get('numberdensity_unit', (1, 'cm**-3'))
+            numberdensity_unit = self.quan(*numberdensity_override)  # numberdensity is never set as attribute
+            density_unit = (1.0 + 4.0 * He_abundance) * mp_cgs * numberdensity_unit
+            mass_unit = density_unit * length_unit**3
+
+        # 2. calculations for velocity
+        if 'time_unit' in self.units_override:
+            # in this case time was supplied
+            velocity_unit = length_unit / self.time_unit
+        else:
+            # other case: velocity was supplied. Fall back to zero (default) if no overrides supplied
+            velocity_unit = getattr(self, 'velocity_unit', None)
+
+        # 3. calculations for pressure and temperature
+        if velocity_unit is None:
+            # velocity and time not given, see if temperature is given. Fall back to one (default) if not
+            temperature_unit = getattr(self, 'temperature_unit', self.quan(1, 'K'))
+            pressure_unit = ((2.0 + 3.0 * He_abundance) * numberdensity_unit * kb_cgs * temperature_unit).to('dyn*cm**-2')
+            velocity_unit = (np.sqrt(pressure_unit / density_unit)).to('cm*s**-1')
```

Comment:
You might want use the `YTQuantity` method `in_cgs()` instead to `to()` here.

Although I'm actually curious, _why_ are you force-converting to cgs ?

Reply:
Because apparently yt does not simplify this by default, if I don't specifically ask for the conversion they remain in terms like `length**1/2` or something like that. This ensures that they play nicely with each other :) And I think I used `to_cgs()` before actually, it probably didn't simplify the units as well so that's why I dropped it

Reply:
Using `in_cgs()` actually worked out nicely, using that now instead